usb: stm32: Get usb domain clock configurable using device tree

erwango opened the pull request:
Similarly to other peripherals, configure USB domain clock using device tree.


This change doesn't impacts to STM32F1 boards that are treated in another PR (#53257).

erwango committed:
drivers: usb: stm32: Remove HSI48 activation related code

Remove this code to clean up the driver.
Clock activation is done from device tree configuration in clock_control
driver.
Users are informed by warning about missing configuration.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers: usb: stm32: Store dt clock information into an array

Use STM32_DT_INST_CLOCKS() to access usb device clock information.
This will help accessing domain clock information from dt nodes.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers: usb: stm32: Configure domain clock using device tree

This change doesn't consider STM32F1 series.

There are 5 categories of STM32 regarding USB domain clock:
A] USB domain clocks directly driven by CK48 clocks
B] USB domain clocks which can select either CK48 or an alternative clock
C] USB domain clocks which can select between MSI or PLL
D] USB domain clocks relying statically on a PLL output
E] USB domain clocks which can be set on 2 different PLL ouputs

For case A], there is nothing to be done. Clk48 was already configured
in a previous change (51c7643295b02ba26ddcd0fd915484114db7101c).
For case B] and C] USB domain clocks will need to be configured in
each board.
For case D] and E] relying on existing dts configuration. Though
there is no easy build time chekc that can be done.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: stm32: Enable USB domain clock when required

For boards configuring USB device, enable or configure matching domain
clock when not done already.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
`DNM` as I still need to treat boards under the following conditions:
```
#if defined(LL_RCC_USB_CLKSOURCE_NONE)
	/* When MSI is configured in PLL mode with a 32.768 kHz clock source,
	 * the MSI frequency can be automatically trimmed by hardware to reach
	 * better than ±0.25% accuracy. In this mode the MSI can feed the USB
	 * device. For now, we only use MSI for USB if not already used as
	 * system clock source.
	 */
```

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -194,88 +196,55 @@ void HAL_PCD_SOFCallback(PCD_HandleTypeDef *hpcd)
-	}
-	/* Force 48 MHz mode */
-	LL_RCC_MSI_EnableRangeSelection();
-	LL_RCC_MSI_SetRange(LL_RCC_MSIRANGE_11);
-	LL_RCC_SetUSBClockSource(LL_RCC_USB_CLKSOURCE_MSI);
-#else
-	if (LL_RCC_PLL_IsReady()) {
-		LL_RCC_SetUSBClockSource(LL_RCC_USB_CLKSOURCE_PLL);
-	} else {
-		LOG_ERR("Unable to set USB clock source to PLL.");
+#warning TODO, configure MSI as USB clck source
+#endif
```
@loicpoulain I'm having hard time to figure out which are the impacted references.
Since you authored these lines, can you tell ?

loicpoulain replied:
What do you mean by references? The impacted boards? Need to double check that, afair trimmed MSI is used when no external crystal (hse) is connected to the microcontroller.

erwango replied:
Ok, thanks. I've tried to clarify things.

jfischer-no commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -45,6 +45,7 @@ LOG_MODULE_REGISTER(usb_dc_stm32);
 #elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_usb)
 #define DT_DRV_COMPAT st_stm32_usb
 #define USB_IRQ_NAME  usb
+
```
Looks like stray change.

FRASTM commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -201,53 +201,42 @@ static int usb_dc_stm32_clock_enable(void)
+
+#if DT_NODE_HAS_COMPAT(DT_NODELABEL(clk_hsi48), fixed_clock)  && !STM32_HSI48_ENABLED
+	/* On these series, HSI48 is available and set by default as USB clok source */
+	/* HSI48 clock not enebaled */
```
/* HSI48 clock not enabled */

erwango replied:
fixed

FRASTM commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -194,88 +196,47 @@ void HAL_PCD_SOFCallback(PCD_HandleTypeDef *hpcd)
-	/* Specially for STM32WB, don't unlock the HSEM to prevent M0 core
-	 * to disable HSI48 clock used for RNG.
-	 */
-	z_stm32_hsem_unlock(CFG_HW_CLK48_CONFIG_SEMID);
```
Is this stm32WB specificity no more valid ?

erwango replied:
Good point, I've added a commit to take this into account in clock_control driver.

erwango committed:
drivers: clock_control: stm32wb: Lock CLK48 hsem

On stm32wb, M0 core may enable and disable CLK48 when using RNG.
Lock related hsem to prevent M0 to disable CLK48 when it doesn't need it
anymore.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
doc: release-notes 3.3: Document STM32 USB clock related changes

Document recent changes about USB clock source configuration.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
> `DNM` as I still need to treat boards under the following conditions:

Tested with disco_l475_iot1.

jfischer-no reviewed (dismissed):
LGTM

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -200,53 +200,42 @@ static int usb_dc_stm32_clock_enable(void)
 		LOG_ERR("clock control device not ready");
 		return -ENODEV;
 	}
-	/*
```
ok, fixed

erwango commented:
`DNM`: F7 boards domain clock configuration have to be checked

erwango committed:
dts: stm32: Describe default domain clock configuration for usb devices

Describe USB default domain clock on all USB devices.
When available select HSI48.
On some series, a default clock my be set by default at start up.
On those series, in order to be able to compute USB clock at runtime,
clock_control driver needs to be aware of configuration and then this
default config has to be described explicitly too.

Default clocks are not enabled though. It is up to board configuration
to provide correct clock configuration (and we should not enabled by
default clocks that would not be required by board configuration).

Note: This change doesn't consider STM32F1/F3 devices, which
have a specific USB clock configuration.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers: usb: stm32: Replace static clock check by a runtime check.

In order to clean up clock related code, replace static build time
clock configuration code by a runtime check.
Since I'm not able to guarantee this check will provide a correct
verdict in 100% of the cases (I'm thinking to ULPI or exotic configurations
like F7 OTG-HS), add a config option to disable it if needed.
It also doesn't apply on F1/F3 series. Keep the build time check for now.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: nucleo f7xx: Increase core speed

STM32F7 allow to run at significantly higher speed that the one configured
on these boards.
Configure clocks to make benefit of full potential.
Additionally, configure PLL_Q output to provide USB compatible 48MHz freq.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: olimex f4 based: Provide a USB compatible clock on PLL_Q

On these boards, adapt clock configuration to output a USB compatible
48MHz on PLL_Q.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: stm32f469i_disco: Tune freq to provide USB compatible freq

In order to be able to provide 48MHz on PLL-Q, slightly decrease
core clock freq.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

gautierg-st commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -261,11 +257,6 @@ static int usb_dc_stm32_clock_enable(void)
 
 #endif /* RCC_HSI48_SUPPORT / LL_RCC_USB_CLKSOURCE_NONE */
```
The comment should be `/* RCC_CFGR_OTGFSPRE / RCC_CFGR_USBPRE */`

erwango replied:
Fixed, Txs

gautierg-st reviewed (changes requested):
Except for the wrong comment, LGTM
